tests/mpi/hp_step-4.cc

```diff
@@ -0,0 +1,352 @@
+
+    MPI_Comm communicator;
+
+    parallel::distributed::Triangulation<dim> triangulation;
+    hp::FECollection<dim>                     fe;
+    hp::DoFHandler<dim>                       dof_handler;
+    hp::QCollection<dim>                      q_collection;
+
+    IndexSet locally_owned_dofs;
+
+    IndexSet locally_relevant_dofs;
+
+    TrilinosWrappers::SparseMatrix system_matrix;
+
+    TrilinosWrappers::MPI::Vector solution, rhs;
+
+    AffineConstraints<double> constraints;
+
+    ConditionalOStream pcout;
+  };
+
+
+
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide()
+      : Function<dim>()
+    {}
+
+    virtual double
+    value(const Point<dim> &p, const unsigned int component = 0) const override;
+  };
+
+
+
+  template <int dim>
+  double
+  RightHandSide<dim>::value(const Point<dim> &p,
+                            const unsigned int /*component*/) const
+  {
+    double return_value = 0.0;
+    for (unsigned int i = 0; i < dim; ++i)
+      return_value += 4.0 * std::pow(p(i), 4.0);
+
+    return return_value;
+  }
+
+
+
+  template <int dim>
+  Step4<dim>::Step4()
+    : communicator(MPI_COMM_WORLD)
+    , triangulation(communicator,
+                    typename Triangulation<dim>::MeshSmoothing(
+                      Triangulation<dim>::smoothing_on_refinement |
+                      Triangulation<dim>::smoothing_on_coarsening))
+    , dof_handler(triangulation)
+    , pcout(std::cout, (Utilities::MPI::this_mpi_process(communicator) == 0))
+  {
+    fe.push_back(FE_Q<dim>(2));
+    fe.push_back(FE_Nothing<dim>());
+
+    q_collection.push_back(QGauss<dim>(fe[0].degree + 1));
+  }
+
+
+
+  template <int dim>
+  void
+  Step4<dim>::make_grid()
+  {
+    GridGenerator::hyper_cube(triangulation, -1, 1);
+    triangulation.refine_global(4);
+
+    for (auto &cell : triangulation.active_cell_iterators())
+      if (cell->is_locally_owned())
+        if ((cell->center()[1] < 0.) && (cell->center()[1] > -0.2))
+          cell->set_refine_flag();
+
+    triangulation.execute_coarsening_and_refinement();
+  }
+
+
+
+  template <int dim>
+  void
+  Step4<dim>::setup_system()
+  {
+    for (auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          Point<dim> center = cell->center();
+
+          if (center[1] >= 0)
+            cell->set_active_fe_index(1);
+
+          else
+            cell->set_active_fe_index(0);
+        }
+    dof_handler.distribute_dofs(fe);
+
+    locally_owned_dofs = dof_handler.locally_owned_dofs();
+    locally_relevant_dofs.clear();
+    DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+    locally_relevant_dofs.compress();
+
+    constraints.clear();
+
+    constraints.reinit(locally_relevant_dofs);
+
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+
+    DoFTools::make_zero_boundary_constraints(dof_handler, constraints);
+
```

Comment:
You are working with `p::d::Triangulation` objects. We still don't know whether we will ever encounter a case in which constraints depend on degrees of freedom of ghost cells, which again are constrained to degrees of freedom of a different subdomains. We check for these 'nested' constraints in other tests. It would be a good idea to add this here, too!

Like this:
https://github.com/dealii/dealii/blob/22c4b6d5ead0a0afcc361bbdb7d46d23e72bbc74/tests/mpi/trilinos_step-27.cc#L232-L244

Reply:
Done!